server/controllers/admin/unit.js

```diff
@@ -0,0 +1,72 @@
+const db = require('../../lib/db');
+const q = require('q');
+
+module.exports.list = list;
+
+/**
+ *
+ * @function list
+ * return the tree
+ * each modules and his pages
+ */
+function list(req, res, next) {
+  const roleUuid = db.bid(req.params.roleUuid);
+
+  // permission(pages) in a module
+  // the column affect will inform is the user has acces to the permission
+  const sql = `
+    SELECT DISTINCT rqt.id,  rqt.name, rqt.key , rqt.parent, rqt.affected FROM (
+      SELECT u.*,  1 as affected
+      FROM unit u
+      JOIN role_unit as ru ON ru.unit_id = u.id
+      WHERE ru.role_uuid = ?
+
+      UNION 
+
+      SELECT  unit.*,  0 as affected
+      FROM unit 
+      WHERE id NOT IN(
+        SELECT  u.id
+        FROM unit u
+        JOIN role_unit as ru ON ru.unit_id = u.id
+        WHERE ru.role_uuid =?
+      )
+    ) as rqt
+  `;
+  // get modules
+
+  // parent nodes
+  const sql1 = `${sql}\n WHERE rqt.parent=0`;
+  // children
+  const sql2 = `${sql}\n WHERE rqt.parent=?`;
+
+  let modules = [];
+  const queries = [];
+
+  db.exec(sql1, [roleUuid, roleUuid]).then((rows) => {
+    modules = rows;
+    rows.forEach((row, idx) => {
```

Comment:
A good rule to remember is that every time you do this kind of statement:
```js
const list = []
array.forEach(x => {
  list.push(x);
}):
return list
```
.. you are actually doing a `map()`.  In the above case, you can do this:
```js
return q.all(rows.map(row => db.exec(sql2, [roleUuid, roleUuid, row.id])));
```